tsl/test/sql/compression_ddl.sql

```diff
@@ -490,3 +490,19 @@ WHERE reltablespace in
 
 DROP TABLE test2 CASCADE;
 DROP TABLESPACE tablespace2;
+
+-- Create a table with a compressed table and then delete the
+-- compressed table and see that the drop of the hypertable does not
+-- generate an error. This scenario can be triggered if an extension
+-- is created with compressed hypertables since the tables are dropped
+-- as part of the drop of the extension.
+CREATE TABLE issue4140("time" timestamptz NOT NULL, device_id int);
+SELECT create_hypertable('issue4140', 'time');
+ALTER TABLE issue4140 SET(timescaledb.compress);
+SELECT format('%I.%I', schema_name, table_name)::regclass AS ctable
+FROM _timescaledb_catalog.hypertable
+WHERE id = (SELECT compressed_hypertable_id FROM _timescaledb_catalog.hypertable WHERE table_name = 'issue4140') \gset
+SELECT timescaledb_pre_restore();
```

Comment:
Why do you need `timescaledb_pre_restore()`?

Reply:
Could you please clarify? Do you mean this scenario cannot happen in other situations, only during an extension update? Is it possible, for instance,  to drop_chunks on a compressed hypertable and get into a bad catalog state?

Reply:
I know it cannot happen through a normal hypertable drop with a compressed hypertable, because I tested that.

It is not relevant for dropping a chunk because the error is triggered because the *compressed hypertable* is dropped before the code reaches the actual lookup of the compressed hypertable.

Reply:
This particular situation happens because `sql_drop` event is used to delete the tables when the extension is dropped.

In this case, the `sql_drop` event is called after the table is actually deleted, so the tables are deleted first and then `ts_timescaledb_process_ddl_event` is called. Once the code to process this delete event is handled, the tables are already gone, so the lookup of the relid fails, triggering the error.

It is entirely possible that running drop chunks on other tables might trigger an error, in particular if internal tables are deleted willy-nilly, but that is out of scope for this issue.